bundle/edu.gemini.ags/src/main/scala/edu/gemini/ags/api/AgsHash.scala

```diff
@@ -0,0 +1,148 @@
+package edu.gemini.ags.api
+
+import edu.gemini.shared.util.immutable.ScalaConverters._
+import edu.gemini.shared.util.immutable.{Option => GemOption}
+import edu.gemini.spModel.ags.AgsStrategyKey
+import edu.gemini.spModel.gemini.flamingos2.Flamingos2
+import edu.gemini.spModel.gemini.gmos.{InstGmosNorth, InstGmosSouth}
+import edu.gemini.spModel.obs.context.ObsContext
+
+import java.time.Instant
+
+import scala.collection.mutable.ListBuffer
+import scala.util.hashing.{MurmurHash3 => M3}
+
+import scala.collection.JavaConverters._
+
+/** Computes an `Int` hash value that corresponds to the set of inputs to the
+  * AGS lookup.  If an observation is changed in some way and yet the hash
+  * algorithm computes the same value, there is no need to re-do the AGS search.
+  */
+object AgsHash {
+
+  /** Calculates the AGS hash for this context, using the current base position
+    * location.
+    *
+    * <em>Warning.</em> This method is not referentially transparent as it may
+    * provide distinct answers for the same input at different times
+    * (particularly for non-sidereal targets).
+    */
+  def hashNow(ctx: ObsContext): Int =
```

Comment:
I expect we won't be calling `hashNow` but rather the other `hash` methods with an explicit time that will be taken from the observing block.  This is the same time that will be used for parallactic angle calculations and (in particular) non-sidereal target coordinates.

Even if you do call `hashNow` though, the hash computation will only differ if the base position changes significantly.  The coordinates are rounded.  For sidereal targets a long time will usually pass before that happens even for the fastest moving targets -- and in fact we aren't yet even considering proper motion etc.  Even a non-sidereal target would have to be moving fast to have a different hash value every millisecond.

At any rate, the hash value is only as valid as our determination of where the target is in the sky so it *should* give us a different answer if the target is in a different location.

This raises a good point though, and I don't know the answer: how much movement should be sufficient to invalidate a hash value?  I wanted milli-arcseconds to count but I'm not sure if that is correct or even if I achieved it with a scale of 8 digits.  I picked 8 because 1 mas is 1/3,600,000 degrees or about 0.00000027778 degrees.  This may well be stupid though.